src/goto-instrument/function_stubs.cpp

```diff
@@ -0,0 +1,264 @@
+/*******************************************************************\
+
+Module: Create function stubs in C from code contracts
+
+Author: Konstantinos Kallas
+
+Date: July 2019
+
+\*******************************************************************/
+
+/// \file
+/// Create function stubs in C from code contracts
+
+#include "function_stubs.h"
+
+#include <fstream>
+
+#include <util/expr_util.h>
+#include <util/fresh_symbol.h>
+#include <util/replace_symbol.h>
+
+#include <goto-programs/remove_skip.h>
+
+#include <analyses/local_may_alias.h>
+
+#include <linking/static_lifetime_init.h>
+
+#include "loop_utils.h"
+
+class function_stubst
+{
+public:
+  function_stubst(
+    symbol_tablet &_symbol_table,
+    goto_functionst &_goto_functions,
+    irep_idt _function_name,
+    messaget &_log)
+    : ns(_symbol_table),
+      symbol_table(_symbol_table),
+      goto_functions(_goto_functions),
+      function_name(_function_name),
+      log(_log),
+      temporary_counter(0)
+  {
+  }
+
+  bool operator()();
+
+protected:
+  namespacet ns;
+  symbol_tablet &symbol_table;
+  goto_functionst &goto_functions;
+  const irep_idt function_name;
+  messaget &log;
+
+  unsigned temporary_counter;
+
+  void stub_function(
+    const irep_idt &function_id,
+    goto_functionst::goto_functiont &goto_function);
+
+  const symbolt &new_tmp_symbol(
+    const typet &type,
+    const source_locationt &source_location,
+    const irep_idt &function_id,
+    const irep_idt &mode);
+};
+
+// It creates a stub for each function based on the associated
+// requires and ensures
+void function_stubst::stub_function(
+  const irep_idt &function_id,
+  goto_functionst::goto_functiont &goto_function)
+{
+  // Get the function requires and ensures
+  const symbolt &f_sym = ns.lookup(function_id);
+  const code_typet &type = to_code_type(f_sym.type);
+
+  const exprt &requires =
+    static_cast<const exprt &>(type.find(ID_C_spec_requires));
+  const exprt &ensures =
+    static_cast<const exprt &>(type.find(ID_C_spec_ensures));
+
+  // Clear the function body
+  goto_function.body.clear();
+
+  // Add the assertion in the function body
+  if(requires.is_not_nil())
+    goto_function.body.add(
+      goto_programt::make_assertion(requires, requires.source_location()));
+
+  // Get the implementation function symbol
+  //
+  // Question: Should I do lookups on the symbol table or should I do
+  // them on the namespace?
+  const symbolt &impl_f_sym =
```

Comment:
Avoid abbreviations and single-char names like these -- `implementation_symbol` is probably good here

Reply:
Add comment to clarify, since the names contradict each other a bit: are you expecting this to be a stub (undefined), or for the user to provide a `__stub__f` function of some sort?

Reply:
Thanks I will add a comment